README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# idmact

<!-- badges: start -->
<!-- badges: end -->

The goal of idmact is to provide an implementation of the [Schiel
(1998)](https://www.act.org/content/dam/act/unsecured/documents/ACT_RR98-01.pdf)
algorithm for interpreting differences between mean ACT assessment
scores.

## Installation

You can install the released version of idmact from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("idmact")
```

You can install the development version of idmact from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mncube/idmact")
```

## Main Algorithm

The idmact package provides tools designed to examine the significance
of differences between mean ACT scale scores over time. The fundamental
algorithm for composite scale scores consists of the following six
steps:

1.  Add one unit to the raw score of one or more subjects for each
    student to derive adjusted raw scores.

2.  Convert adjusted raw scores to adjusted scale scores using the
    form’s raw score to scale score map. Note that perfect raw scores
    are always converted to the maximum allowable scale score,
    irrespective of the adjustment in step one.

3.  For each student, sum the adjusted scale scores across all subject
    areas, divide this sum by the number of subject areas, and round to
    the nearest integer. This produces each student’s adjusted composite
    scale score.

4.  Calculate the adjusted mean composite scale score across all
    students (m_adj).

5.  Calculate the unadjusted mean composite scale score across all
    observations (m_unadj).

6.  Compute the difference between the adjusted and unadjusted mean
    composite scale scores to get the delta composite: deltac = m_adj -
    m_unadj

While this algorithm was initially developed to address the challenges
in interpreting differences between mean ACT scale scores, it can also
be beneficial in other contexts. These may include situations where
different forms of an assessment have varying raw score to scale score
maps, particularly when the relationship between raw and scale scores is
complex and/or proprietary.

## Single Subject Example

In the subsequent example, we’ll use the algorithm described above to
interpret differences between two forms of a hypothetical assessment.
Both forms share the same range of raw scores and scale scores. However,
the conversion map from raw scores to scale scores differs between the
two assessments. To keep this example straightforward, we’ll simulate a
single data set of raw scores and then use the algorithm to compare mean
differences in scale scores between the two assessments. This initial
example will concentrate on a single subject area, hence step 3 of the
algorithm will not be included.

### Generate Data and Maps

``` r
library(idmact)

# Create 100 raw scores
set.seed(279)
raw_scores <- as.list(sample(1:100, 100, replace = TRUE))

# Map between raw scores and scale scores for each form

## Each form has scale scores ranging from 1 to 12
map_scale <- c(1:12)

## Each assessment has raw scores ranging from 1 - 100
map_raw_formA <- list(1:5, 6:20, 21:25, 26:40, 41:45, 46:50, 51:55,
                   56:75, 76:80, 81:85, 86:90, 91:100)

map_raw_formB <- list(1:10, 11:20, 21:30, 31:40, 41:50, 51:55, 56:65,
                   66:75, 76:85, 86:90, 91:95, 96:100)
```

### Format Map

In the raw score to scale score map presented above, vectors/lists were
utilized to efficiently map each of the 100 raw scores to each of the 12
scale scores. Use the map_elongate function to expand the raw and scale
sections of the map into a format where each portion is a list of 100.
This is the map format required for the idmact_subj function, which
implements the subject-level algorithm.

``` r
formA <- map_elongate(map_raw = map_raw_formA,
                      map_scale = map_scale)

formB <- map_elongate(map_raw = map_raw_formB,
                      map_scale = map_scale)
```

### Run Subject Level Algorithm

Utilize the raw data and the maps stored in formA and formB to calculate
and compare the subject-level delta (deltas). In the algorithm below,
each raw score is increased by 1 using the default value of the inc
parameter.

``` r
resA <- idmact_subj(raw = raw_scores,
                    map_raw = formA$map_raw,
                    map_scale = formA$map_scale)


resB <- idmact_subj(raw = raw_scores,
                    map_raw = formB$map_raw,
                    map_scale = formB$map_scale)
```

### Compare Results

``` r
cat("Form A subject level delta:", resA$deltas, "\n")
#> Form A subject level delta: 0.11
cat("Form B subject level delta:", resB$deltas)
#> Form B subject level delta: 0.1
```

In this section, the function idmact_subj is used to calculate the
‘delta’ for each form (A and B). The ‘delta’ is the difference between
the mean adjusted scale score and the mean unadjusted scale score. This
‘delta’ value provides an estimate of how much the mean scale score
would increase if every student were to answer one additional item
correctly on the test.

In the provided example:

For Form A, the subject level delta is 0.11, meaning the mean scale
score is expected to increase by 0.11 if every student answers one
additional item correctly.

For Form B, the subject level delta is 0.1, suggesting that the mean
scale score would increase by 0.1 under the same conditions.

The results indicate that Form A is more responsive to increases in raw
scores, as a one unit increase in raw score leads to a larger increase
in the mean scale score for Form A compared to Form B.

## Composite Example

The idmact_comp function can be used to calculate the delta for
composite scores (deltac). In the example below, raw scores for an
additional subject area (‘s2’) will be created and combined with the
data from the previous example to demonstrate idmact_comp.

### Generate Data and Maps

``` r
# Create 100 raw scores
set.seed(250)
raw_scores_s2 <- as.list(sample(1:100, 100, replace = TRUE))

# Subject two will use the same ranges for raw scores and scale scores as in
# the previous example, but the map will be slightly different. 
map_raw_formA_s2 <- list(1:10, 11:25, 26:30, 31:40, 41:45, 46:50, 51:60,
                         61:75, 76:80, 81:85, 86:90, 91:100)

map_raw_formB_s2 <- list(1:10, 11:16, 17:25, 26:35, 36:45, 46:55, 56:60,
                         61:75, 76:85, 86:90, 91:95, 96:100)

formA_s2 <- map_elongate(map_raw = map_raw_formA_s2,
                         map_scale = map_scale)

formB_s2 <- map_elongate(map_raw = map_raw_formB_s2,
                         map_scale = map_scale)
```

### Run Composite Level Algorithm

In the algorithm below, each raw score for each subject is incremented
by 1.

``` r
resA_comp <- idmact_comp(raw = list(raw_scores, raw_scores_s2),
                         inc = list(1, 1),
                         map_raw = list(formA$map_raw, formA_s2$map_raw),
                         map_scale = list(formA$map_scale, formA_s2$map_scale))

resB_comp <- idmact_comp(raw = list(raw_scores, raw_scores_s2),
                         inc = list(1, 1),
                         map_raw = list(formB$map_raw, formB_s2$map_raw),
                         map_scale = list(formB$map_scale, formB_s2$map_scale))
```

### Compare Composite Results

``` r
cat("Form A composite level delta:", resA_comp$composite_results$deltac, "\n")
#> Form A composite level delta: 0.08
cat("Form B composite level delta:", resB_comp$composite_results$deltac)
#> Form B composite level delta: 0.08
```

In the composite example, two subjects are considered instead of one,
and the idmact_comp function is used to calculate the composite level
delta (deltac). The composite level delta is calculated in a similar way
to the subject level delta, but it considers the total adjusted and
unadjusted scale scores across all subjects, rather than just one.

In the provided example:

For both Form A and Form B, the composite level delta is 0.08. This
means that if every student were to answer one additional item correctly
on each form, the mean composite scale score is expected to increase by
0.08.

This suggests that, when considering multiple subjects, both Form A and
Form B respond similarly to increases in raw scores.

### See Subject Area 2 Results

``` r
cat("Form A subject area 2 delta:", resA_comp$subject_results[[2]]$deltas, "\n")
#> Form A subject area 2 delta: 0.05
cat("Form B subject area 2 delta:", resB_comp$subject_results[[2]]$deltas)
#> Form B subject area 2 delta: 0.08
```

This section presents the ‘delta’ for the second subject area
specifically. The ‘delta’ for the second subject area is calculated in
the same way as the subject level delta mentioned in the first section,
but it only considers the scores for the second subject.

In the provided example:

For Form A, the subject area 2 delta is 0.05, suggesting that the mean
scale score for the second subject would increase by 0.05 if every
student answered one additional item correctly.

For Form B, the subject area 2 delta is 0.08, meaning that the mean
scale score for the second subject would increase by 0.08 under the same
conditions.

These results indicate that, for the second subject area specifically,
Form B is more responsive to increases in raw scores.


NEWS.md
# idmact 1.0.1.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(idmact)

test_check("idmact")


tests/testthat/test-idmact_comp.R
test_that("idmact_comp works", {
  # Test all list inputs

  # Inputs
  raw = list(list(1, 2, 3, 4, 5), list(1, 1, 1, 1, 1))
  inc = list(1 , 1)
  map_raw = list(list(1, 2, 3, 4, 5))
  map_scale = list(list(20, 21, 22, 23, 24))
    # # Debugging inputs
    # df = NULL
    # df_map = NULL
    # mcent = "mean"
    # na.rm.mcent = TRUE

  comp_mean <- idmact_comp(raw = raw,
                           inc = inc,
                           map_raw = map_raw,
                           map_scale = map_scale)

  expect_equal(comp_mean$composite_results$deltac, 0.8)

  # Test raw scores in df map in lists

  #Inputs
  df <- data.frame(Id = c(1:20),
                   Score1 = rep(11:15, 4),
                   Score2 = rep(11:14, 5))
  raw = list("Score1", "Score2")
  inc = list(1 , 1)
  map_raw <- list(list(11, 12, 13, 14, 15))
  map_scale <- list(list(21, 22, 23, 24, 25))


  comp_dfraws_mean <- idmact_comp(df = df,
                                  raw = raw,
                                  inc = inc,
                                  map_raw = map_raw,
                                  map_scale = map_scale)

  expect_equal(comp_dfraws_mean$composite_results$deltac, 0.9)

  # Test raw scores and map in same df

  df <- data.frame(Id = c(1:20),
                   Score1 = rep(11:15, 4),
                   Score2 = rep(12:15, 5),
                   rawmap = rep(11:15, 4),
                   scalemap = rep(21:25, 4))
  raw = list("Score1", "Score2")
  inc = list(1 , 1)
  map_raw <- list("rawmap")
  map_scale <- list("scalemap")

  comp_samedf_mean <- idmact_comp(df = df,
                                  raw = raw,
                                  inc = inc,
                                  map_raw = map_raw,
                                  map_scale = map_scale)

  expect_equal(comp_samedf_mean$composite_results$deltac, 0.75)

  # Test raw scores and map in seperate dfs

  df <- data.frame(Id = c(1:20),
                   Score1 = rep(11:15, 4),
                   Score2 = rep(12:15, 5))
  df_map <- data.frame(rawmap = rep(11:15, 4),
                       scalemap = rep(21:25, 4))
  raw = list("Score1", "Score2")
  inc = list(1 , 1)
  map_raw <- list("rawmap")
  map_scale <- list("scalemap")

  comp_samedf_mean <- idmact_comp(df = df,
                                  df_map = df_map,
                                  raw = raw,
                                  inc = inc,
                                  map_raw = map_raw,
                                  map_scale = map_scale)

  expect_equal(comp_samedf_mean$composite_results$deltac, 0.75)

})


test_that("idmact_comp sanity check: reverse maps with symmetric constant raw
          scores gives a deltac of 0", {

            # Test 2s and 4s
            raw1 = list(2,2,2)
            map_raw1 = list(1,2, 3, 4, 5)
            map_scale1 = list(10, 20, 30, 40, 50)

            raw2 = list(4,4,4)
            map_raw2 = list(1,2, 3, 4, 5)
            map_scale2 = list(50, 40, 30, 20, 10)

            inc = list(1 , 1)

            expect_equal(idmact_comp(raw = list(raw1, raw2),
                                     map_raw = list(map_raw1, map_raw2),
                                     map_scale = list(map_scale1, map_scale2),
                                     inc = inc)$composite_results$deltac, 0)
            })

test_that("README Composite Example", {
  # Create 100 raw scores
  set.seed(279)
  raw_scores <- as.list(sample(1:100, 100, replace = TRUE))

  # Map between raw scores and scale scores for each form

  ## Each form has scale scores ranging from 1 to 12
  map_scale <- c(1:12)

  ## Each assessment has raw scores ranging from 1 - 100
  map_raw_formA <- list(1:5, 6:20, 21:25, 26:40, 41:45, 46:50, 51:55,
                        56:75, 76:80, 81:85, 86:90, 91:100)

  map_raw_formB <- list(1:10, 11:20, 21:30, 31:40, 41:50, 51:55, 56:65,
                        66:75, 76:85, 86:90, 91:95, 96:100)

  formA <- map_elongate(map_raw = map_raw_formA,
                        map_scale = map_scale)

  formB <- map_elongate(map_raw = map_raw_formB,
                        map_scale = map_scale)

  resA <- idmact_subj(raw = raw_scores,
                      map_raw = formA$map_raw,
                      map_scale = formA$map_scale)


  resB <- idmact_subj(raw = raw_scores,
                      map_raw = formB$map_raw,
                      map_scale = formB$map_scale)

  # Create 100 raw scores
  set.seed(250)
  raw_scores_s2 <- as.list(sample(1:100, 100, replace = TRUE))

  # Subject to will use the same ranges for raw scores and scale scores as was used
  # in the previous example, but the map will be slightly different.
  map_raw_formA_s2 <- list(1:10, 11:25, 26:30, 31:40, 41:45, 46:50, 51:60,
                           61:75, 76:80, 81:85, 86:90, 91:100)

  map_raw_formB_s2 <- list(1:10, 11:16, 17:25, 26:35, 36:45, 46:55, 56:60,
                           61:75, 76:85, 86:90, 91:95, 96:100)

  formA_s2 <- map_elongate(map_raw = map_raw_formA_s2,
                           map_scale = map_scale)

  formB_s2 <- map_elongate(map_raw = map_raw_formB_s2,
                           map_scale = map_scale)

  resA_comp <- idmact_comp(raw = list(raw_scores, raw_scores_s2),
                           inc = list(1, 1),
                           map_raw = list(formA$map_raw, formA_s2$map_raw),
                           map_scale = list(formA$map_scale, formA_s2$map_scale))

  resB_comp <- idmact_comp(raw = list(raw_scores, raw_scores_s2),
                           inc = list(1, 1),
                           map_raw = list(formB$map_raw, formB_s2$map_raw),
                           map_scale = list(formB$map_scale, formB_s2$map_scale))

  resA_s2 <- idmact_subj(raw = raw_scores_s2,
                      map_raw = formA_s2$map_raw,
                      map_scale = formA_s2$map_scale)


  resB_s2 <- idmact_subj(raw = raw_scores_s2,
                      map_raw = formB_s2$map_raw,
                      map_scale = formB_s2$map_scale)

  # 3.  Sum the adjusted scale scores for each subject area, divide this sum by
  #the number of subject areas, and round to the nearest integer in order to
  #obtain each observation/examinee's adjusted composite scale score.
  mean_scale_unadj_A <- round((unlist(resA$scale$unadj) + unlist(resA_s2$scale$unadj))/2)
  mean_scale_adj_A <- round((unlist(resA$scale$adj) + unlist(resA_s2$scale$adj))/2)
  mean_scale_unadj_B <- round((unlist(resB$scale$unadj) + unlist(resB_s2$scale$unadj))/2)
  mean_scale_adj_B <- round((unlist(resB$scale$adj) + unlist(resB_s2$scale$adj))/2)

  # 4.  Calculate the adjusted mean composite scale score across all observations
  # (m_adj).
  m_adj_A <- mean(mean_scale_adj_A)
  m_adj_B <- mean(mean_scale_adj_B)

  # 5.  Calculate the unadjusted mean composite scale score across all observations
  # (m_unadj).
  m_unadj_A <- mean(mean_scale_unadj_A)
  m_unadj_B <- mean(mean_scale_unadj_B)

  # 6.  Compute the difference between the adjusted and unadjusted mean composite
  # scale scores to obtain delta composite: deltac = m_adj - m_unadj
  deltacA <- m_adj_A - m_unadj_A
  deltacB <- m_adj_B - m_unadj_B

  # Compare mostly-hand computation to idmact_comp computation
  expect_equal(resA_comp$composite_results$deltac, deltacA)
  expect_equal(resB_comp$composite_results$deltac, deltacB)

})


test_that("idmact_comp's mcen_obs works as expected", {
  # Test all list inputs

  # Inputs
  raw = list(list(1, 2, 3, 4, 5), list(1, 1, 1, 1, 1))
  inc = list(1 , 1)
  map_raw = list(list(1, 2, 3, 4, 5))
  map_scale = list(list(20, 21, 22, 23, 24))

  comp_mean1 <- idmact_comp(raw = raw,
                           inc = inc,
                           map_raw = map_raw,
                           map_scale = map_scale)

  comp_mean2 <- idmact_comp(raw = raw,
                            inc = inc,
                            map_raw = map_raw,
                            map_scale = map_scale,
                            mcent_obs = function(x) round(mean(x)))



  expect_equal(comp_mean1$composite_results$deltac,
               comp_mean2$composite_results$deltac)
  })


tests/testthat/test-idmact_subj.R
test_that("idmact_subj works", {

  # Test list input
  raw = list(1, 2, 3, 4, 5)
  map_raw = list(1, 2, 3, 4, 5)
  map_scale = list(20, 21, 22, 23, 24)

  expect_equal(idmact_subj(raw = raw,
                           map_raw = map_raw,
                           map_scale = map_scale)$deltas, 0.8)

  # Test data frame scores and list maps
  df <- data.frame(Id = c(1:20),
                   Score = rep(11:15, 4))


  expect_equal(idmact_subj(df = df, raw = "Score",
                           map_raw = list(11, 12, 13, 14, 15),
                           map_scale = list(21, 22, 23, 24, 25))$deltas,
               0.8)

  # Test scores and maps in one data frame
  df <- data.frame(Id = c(1:20),
                   Score = rep(11:15, 4),
                   rawmap = rep(11:15, 4),
                   scalemap = rep(21:25, 4))

  expect_equal(idmact_subj(df = df, raw = "Score",
                            map_raw = "rawmap",
                            map_scale = "scalemap")$deltas, 0.8)

  # Test scores and maps in separate data frames
  df <- data.frame(Id = c(1:20),
                   Score = rep(11:15, 4))
  df_map <- data.frame(rawmap = rep(11:15, 4),
                       scalemap = rep(21:25, 4))
  expect_equal(idmact_subj(df = df, df_map = df_map, raw = "Score",
                           map_raw = "rawmap",
                           map_scale = "scalemap")$deltas, 0.8)

  # Test beta when measure of central tendency is median
  raw = list(1, 2, 3, 4, 5)
  map_raw = list(1, 2, 3, 4, 5)
  map_scale = list(20, 21, 22, 23, 24)

  expect_equal(idmact_subj(raw = raw,
                           map_raw = map_raw,
                           map_scale = map_scale,
                           mcent = "median")$deltas, 1)

  # Test missing data
  raw = list(1, 2, 3, NA, 5)
  map_raw = list(1, 2, 3, 4, 5)
  map_scale = list(20, 21, 22, 23, 24)

  expect_equal(idmact_subj(raw = raw,
                           map_raw = map_raw,
                           map_scale = map_scale,
                           mcent = function(x)median(x, na.rm = TRUE))$deltas, 1)


})


test_that("idmact_subj sanity check: small hand calculation", {

  # List computations
  raw = list(2,2,2)
  map_raw = list(1,2,3)
  map_scale = list(10, 20, 30)

  expect_equal(idmact_subj(raw = raw,
                           map_raw = map_raw,
                           map_scale = map_scale)$deltas, 10)

  # One data frame computation
  df <- data.frame(raw = c(2,2,2),
                   rawmap = c(1,2,3),
                   scalemap = c(10, 20, 30))

  expect_equal(idmact_subj(df = df,
                           raw = "raw",
                           map_raw = "rawmap",
                           map_scale = "scalemap")$deltas, 10)

  # Two data frame computation
  df <- data.frame(raw = c(2,2,2))
  df_map <- data.frame(rawmap = c(1,2,3),
                       scalemap = c(10, 20, 30))

  expect_equal(idmact_subj(df = df,
                           df_map = df_map,
                           raw = "raw",
                           map_raw = "rawmap",
                           map_scale = "scalemap")$deltas, 10)


})


test_that("idmact_subj sanity check: reverse maps give equal magnitude opposite
          delta for symmetric constant raw scores", {

            # Test 2s and 4s
            raw1 = list(2,2,2)
            map_raw1 = list(1,2, 3, 4, 5)
            map_scale1 = list(10, 20, 30, 40, 50)

            raw2 = list(4,4,4)
            map_raw2 = list(1,2, 3, 4, 5)
            map_scale2 = list(50, 40, 30, 20, 10)

            expect_equal(idmact_subj(raw = raw1,
                                     map_raw = map_raw1,
                                     map_scale = map_scale1)$deltas,
                         -idmact_subj(raw = raw2,
                                      map_raw = map_raw2,
                                      map_scale = map_scale2)$deltas)
})


test_that("README Single Subject Example", {
  # Create 100 raw scores
  set.seed(279)
  raw_scores <- as.list(sample(1:100, 100, replace = TRUE))

  # Map between raw scores and scale scores for each form

  ## Each form has scale scores ranging from 1 to 12
  map_scale <- c(1:12)

  ## Each assessment has raw scores ranging from 1 - 100
  map_raw_formA <- list(1:5, 6:20, 21:25, 26:40, 41:45, 46:50, 51:55,
                        56:75, 76:80, 81:85, 86:90, 91:100)

  map_raw_formB <- list(1:10, 11:20, 21:30, 31:40, 41:50, 51:55, 56:65,
                        66:75, 76:85, 86:90, 91:95, 96:100)

  formA <- map_elongate(map_raw = map_raw_formA,
                        map_scale = map_scale)

  formB <- map_elongate(map_raw = map_raw_formB,
                        map_scale = map_scale)

  resA <- idmact_subj(raw = raw_scores,
                      map_raw = formA$map_raw,
                      map_scale = formA$map_scale)


  resB <- idmact_subj(raw = raw_scores,
                      map_raw = formB$map_raw,
                      map_scale = formB$map_scale)

  expect_equal(resA$m_scale$adj - resA$m_scale$unadj,
               resA$deltas)

  expect_equal(resB$m_scale$adj - resB$m_scale$unadj,
               resB$deltas)
})


test_that("README Single Subject Example does not throw error when number of
          raw scores does not equal the map length", {
  # Create 100 raw scores
  set.seed(279)
  raw_scores <- as.list(sample(1:100, 233, replace = TRUE))

  # Map between raw scores and scale scores for each form

  ## Each form has scale scores ranging from 1 to 12
  map_scale <- c(1:12)

  ## Each assessment has raw scores ranging from 1 - 100
  map_raw_formA <- list(1:5, 6:20, 21:25, 26:40, 41:45, 46:50, 51:55,
                        56:75, 76:80, 81:85, 86:90, 91:100)

  map_raw_formB <- list(1:10, 11:20, 21:30, 31:40, 41:50, 51:55, 56:65,
                        66:75, 76:85, 86:90, 91:95, 96:100)

  formA <- map_elongate(map_raw = map_raw_formA,
                        map_scale = map_scale)

  formB <- map_elongate(map_raw = map_raw_formB,
                        map_scale = map_scale)

  resA <- idmact_subj(raw = raw_scores,
                      map_raw = formA$map_raw,
                      map_scale = formA$map_scale)


  resB <- idmact_subj(raw = raw_scores,
                      map_raw = formB$map_raw,
                      map_scale = formB$map_scale)

  expect_equal(round(resA$deltas, 3),
               round(0.0944206,3))

  expect_equal(round(resB$deltas, 3),
               round(0.1030043, 3))
})


test_that("Additional-Workflows Example with df and df_map does not throw error
when number of raw scores does not equal the map length", {
  # Create 100 raw scores
  set.seed(279)
  raw_scores <- as.list(sample(1:100, 233, replace = TRUE))

  # Map between raw scores and scale scores for each form

  ## Each form has scale scores ranging from 1 to 12
  map_scale <- c(1:12)

  # First make sure the information in map_raw_formA and map_raw_formB is the same
  # length as the information in map_scale
  map_raw_formA <- as.character(alist(1-5, 6-20, 21-25, 26-40, 41-45, 46-50, 51-55,
                                      56-75, 76-80, 81-85, 86-90, 91-100))

  map_raw_formB <- as.character(alist(1-10, 11-20, 21-30, 31-40, 41-50, 51-55, 56-65,
                                      66-75, 76-85, 86-90, 91-95, 96-100))

  # Then form the df_raw data frames
  df_raw_A <- data.frame(raw = map_raw_formA,
                         scale = map_scale)
  df_raw_A <- map_elongate_df(df_raw_A, "raw", "scale")

  df_raw_B <- data.frame(raw = map_raw_formB,
                         scale = map_scale)
  df_raw_B <- map_elongate_df(df_raw_B, "raw", "scale")

  # Now the df data frame only needs to hold the raw scores
  df <- data.frame(scores = unlist(raw_scores))

  resA <- idmact_subj(df = df,
                      df_map = df_raw_A,
                      raw = "scores",
                      map_raw = "raw",
                      map_scale = "scale")

  resB <- idmact_subj(df = df,
                      df_map = df_raw_B,
                      raw = "scores",
                      map_raw = "raw",
                      map_scale = "scale")

  expect_equal(round(resA$deltas, 3),
                         round(0.0944206,3))
  expect_equal(round(resB$deltas, 3),
                         round(0.1030043, 3))
  })




R/idmact_comp.R
#' Interpreting Differences in Mean ACT Scores at the Composite Level
#'
#' The idmact_comp() function calculates and interprets differences in ACT composite
#' scores. The function operates in the following steps:
#' 1. Increment raw scores for one or more subjects for each student to obtain
#' adjusted raw scores.
#' 2. Map adjusted raw scores to adjusted scale scores using the form's raw score
#' to scale score map.
#' 3. Obtain each examinee's adjusted composite scale score by averaging the
#' adjusted scale scores across subjects.
#' 4. Calculate the adjusted and unadjusted mean composite scale scores across
#' all observations.
#' 5. Compute the difference between the adjusted and unadjusted mean composite
#' scale scores.
#'
#' By default, the function parameters align with the method presented in Schiel
#' (1998). However, you can specify arbitrary anonymous functions for different
#' implementations.
#'
#' @param df A data frame containing raw scores (optional). If provided, the 'raw'
#' parameter should contain column names from this data frame.
#' @param df_map A data frame mapping raw scores to scale scores.
#' @param raw A list of raw scores for each subject, or column names from 'df'
#' where raw scores are stored.
#' @param inc A value or function used to increment raw scores for adjusted score
#' calculation. This can be a single value or a list of values for each subject.
#' @param map_raw Column names from 'df' or 'df_map' representing the domain of
#' raw scores, or a list of such domains.
#' @param map_scale Column names from 'df' or 'df_map' representing the range of
#' scale scores, or a list of such ranges.
#' @param mcent_subj A function summarizing scale scores at the subject level
#' (default is mean with NA removal).
#' @param mcent_obs A function summarizing scale scores at the examinee level
#' (default is round(mean)).
#' @param mcent_comp A function summarizing composite level scale scores
#' (default is mean with NA removal).
#' @param na.rm.max A boolean indicating whether to remove NA values when computing
#' maximum raw and scale values in the mapping.
#'
#' @return A list containing composite and subject level results.
#' "composite_results" includes the difference between the adjusted and unadjusted
#' mean composite scale scores (deltac), the mean adjusted and unadjusted composite
#' scale scores (mscale), and a list of individual adjusted and unadjusted composite
#' scale scores (scale). "subject_results" includes the outcomes from idmact_subj
#' for each subject.
#'
#' @export
#'
#' @references
#' Schiel, J. C. (1998). Interpreting differences in ACT composite scores
#' (ACT Research Report Series 98-1). ACT, Inc. URL:
#' https://www.act.org/content/dam/act/unsecured/documents/ACT_RR98-01.pdf
#'
#' @examples
#' # Example 1: Using df and df_map
#' df <- data.frame(raw1 = c(1, 2, 3), raw2 = c(1, 1, 1))
#' df_map <- data.frame(map_raw1 = c(1, 2, 3),
#'                      map_scale1 = c(20, 21, 22),
#'                      map_raw2 = c(1, 1, 1),
#'                      map_scale2 = c(20, 20, 20))
#' comp_mean <- idmact_comp(df = df,
#'                          df_map = df_map,
#'                          raw = c("raw1", "raw2"),
#'                          inc = 1,
#'                          map_raw = c("map_raw1", "map_raw2"),
#'                          map_scale = c("map_scale1", "map_scale2"))
#'
#' # Example 2: Using lists
#' raw = list(list(1, 2, 3, 4, 5), list(1, 1, 1, 1, 1))
#' inc = list(1 , 1)
#' map_raw = list(list(1, 2, 3, 4, 5))
#' map_scale = list(list(20, 21, 22, 23, 24))
#' comp_mean <- idmact_comp(raw = raw,
#'                          inc = inc,
#'                          map_raw = map_raw,
#'                          map_scale = map_scale)
idmact_comp <- function(df = NULL, df_map = NULL, raw,
                        inc, map_raw, map_scale,
                        mcent_subj = function(x) mean(x, na.rm = TRUE),
                        mcent_obs = function(x) round(sum(x) / length(x)),
                        mcent_comp = function(x) mean(x, na.rm = TRUE),
                        na.rm.max = TRUE){

  # Find the maximum length among raw, inc, map_raw, and map_scale
  max_len <- max(length(raw), length(inc), length(map_raw), length(map_scale))

  # Check and recycle parameters if necessary
  recycle_param <- function(param, max_len) {
    if (length(param) < max_len) {
      param <- rep(param, length.out = max_len)
    }
    return(param)
  }

  raw <- recycle_param(raw, max_len)
  inc <- recycle_param(inc, max_len)
  map_raw <- recycle_param(map_raw, max_len)
  map_scale <- recycle_param(map_scale, max_len)

  # Initialize lists to store results for each subject
  subjr <- vector(mode = "list", length = max_len)
  adj_scale_scores <- vector(mode = "list", length = max_len)
  unadj_scale_scores <- vector(mode = "list", length = max_len)

  # Iterate over subjects
  for (subj in seq_len(max_len)) {

    # Apply idmact_subj for each subject
    subj_results <- idmact_subj(df, df_map, raw = raw[[subj]],
                                inc = inc[[subj]], map_raw = map_raw[[subj]],
                                map_scale = map_scale[[subj]],
                                mcent_subj = mcent_subj,
                                na.rm.max = na.rm.max)

    # Store results for each subject
    subjr[[subj]] <- subj_results
    adj_scale_scores[[subj]] <- subj_results$scale$adj
    unadj_scale_scores[[subj]] <- subj_results$scale$unadj
  }

  # Calculate composite score for each examinee
    adj_comp_scores <- apply(sapply(adj_scale_scores, unlist), 1, mcent_obs)
    unadj_comp_scores <- apply(sapply(unadj_scale_scores, unlist), 1, mcent_obs)

  # Calculate the adjusted and unadjusted mcent ACT Composite scale score
  m_adj_comp <- do.call(mcent_comp, list(adj_comp_scores))
  m_unadj_comp <- do.call(mcent_comp, list(unadj_comp_scores))

  # Calculate the difference between the adjusted and unadjusted mcent Composite scores
  delta_comp <- m_adj_comp - m_unadj_comp

  # Output object
  out <- list("composite_results" = list("deltac" = delta_comp,
                                         "mscale" = list("adj" = m_adj_comp,
                                                         "unadj" = m_unadj_comp),
                                         "scale" = list("adj" = adj_comp_scores,
                                                        "unadj" = unadj_comp_scores)),
              "subject_results" = subjr)

  # Return output
  return(out)
}

